### PR TITLE
Switch the ID to small instance wrapping the long and caching the Strings

### PR DESCRIPTION
**Pull request by bantonsson**
Hacked together this to measure it on the `petclinic` benchmark. Based on #1458. The instance size of a `DDId` is 32 bytes on HotSpot 64bit with COOPS.

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/DatadogHttpCodec.java [hunk 1]
```diff
@@ -32,8 +32,8 @@ private DatadogHttpCodec() {
     public <C> void inject(
         final DDSpanContext context, final C carrier, final AgentPropagation.Setter<C> setter) {
 
-      setter.set(carrier, TRACE_ID_KEY, context.getTraceId().toString());
-      setter.set(carrier, SPAN_ID_KEY, context.getSpanId().toString());
+      setter.set(carrier, TRACE_ID_KEY, String.valueOf(context.getTraceId()));
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/core/propagation/DatadogHttpCodec.java:35]: Again, must be unsigned decimal encoding.

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/HttpCodec.java [hunk 2]
```diff
@@ -114,16 +113,15 @@ public CompoundExtractor(final List<Extractor> extractors) {
    * @throws IllegalArgumentException if value cannot be converted to integer or doesn't conform to
    *     required boundaries
    */
-  static BigInteger validateUInt64BitsID(final String value, final int radix)
+  static long validateUInt64BitsID(final String value, final int radix)
       throws IllegalArgumentException {
-    final BigInteger parsedValue = new StringCachingBigInteger(value, radix);
+    final BigInteger parsedValue = new BigInteger(value, radix);
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/core/propagation/HttpCodec.java:118]: Once we have unsigned parsing, we can parse, not validate here. The JDK8 methods are not `@HotSpotIntrinsicCandidate`, so there is no magic going on, but seem to be many times faster than parsing via `BigInteger`. So porting those routines might make a small stab at the overhead here.

File: dd-trace-core/src/main/java/datadog/trace/core/serialization/FormatWriter.java [hunk 3]
```diff
@@ -125,9 +125,9 @@ public void writeDDSpan(final DDSpan span, final DEST destination) throws IOExce
     /* 1  */ writeString(SERVICE, span.getServiceName(), destination);
     /* 2  */ writeString(NAME, span.getOperationName(), destination);
     /* 3  */ writeString(RESOURCE, span.getResourceName(), destination);
-    /* 4  */ writeBigInteger(TRACE_ID, span.getTraceId(), destination);
-    /* 5  */ writeBigInteger(SPAN_ID, span.getSpanId(), destination);
-    /* 6  */ writeBigInteger(PARENT_ID, span.getParentId(), destination);
+    /* 4  */ writeLong(TRACE_ID, span.getTraceId().toLong(), destination);
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/core/serialization/FormatWriter.java:128]: This is a performance win.

File: dd-trace-core/src/main/java/datadog/trace/core/DDId.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  private final long id;
+  private String str;
+  private String hex;
+
+  private DDId(long id) {
+    this.id = id;
+  }
+
+  public DDId() {
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/core/DDId.java:21]: It also needs to be labeled with its radix because some codecs use hex trace ids.

**Comment by bantonsson** [comment]
Superseded by #1506

**Pull request closed without merging**